### PR TITLE
Set background color on lightning

### LINKED ISSUE #1243: Level "Shocking" is a bit unfair w/o sound
**SuperTux version:** 0.6.1 beta 1
**System information:** GNU/Linux

##### The problem
In the level Shocking from the Forest World, the water occassionally turns deadly because of lightning.

There is a thunder sound played shortly before the water hurts you.

But what if sound is turned off? Then there is no such warning. That's a problem.

### PR DESCRIPTION
**Pull request by tobbi**
Fixes #1243

### CODE REVIEW

**Comment by tobbi** [comment]
https://github.com/SuperTux/supertux/assets/583169/20867e9b-7244-4ccc-b9fb-a32b7b13845e

**Comment by tobbi** [comment]
https://github.com/SuperTux/supertux/assets/583169/59f2b905-d870-4d70-985a-aca4a1151211

**Comment by Rusty-Box** [comment]
Not sure how far we can go in terms of brighting up the screen, but the flash should be very bright! Perhaps working with something on an "additive layer" would help out. I also think that it looks much nicer if the flash appears immideatly upon lighting strike but then fades out rather than dissappearing instantly again.

**Comment by Rusty-Box** [comment]
Ideally would be to having something like this:

lightning_before_after

**Comment by tobbi** [comment]
https://cdn.discordapp.com/attachments/1033726844660351026/1195497037022646393/2024-01-12_23-37-31.mp4?ex=65b43460&is=65a1bf60&hm=b6bf7468679146e4310aab59ec96144d00245653cb7985e0d269b30d8295c7ed&

**Comment by Vankata453** [comment]
Personally I find the thunder too bright, making it uneasy on the eyes. Would there be an option to turn this off?

File: src/object/thunderstorm.cpp [hunk 1]
```diff
@@ -141,6 +153,10 @@ Thunderstorm::lightning()
   if (!m_strike_script.empty()) {
 	  Sector::get().run_script(m_strike_script, "strike-script");
   }
+
+  change_background_colors(true);
+  Sector::current()->get_singleton_by_type<AmbientLight>().set_ambient_light(Color::LIGHTNING_HIT_COLOR);
+  layer = std::numeric_limits<int>::max();
```
Comment (MatusGuy) [inline src/object/thunderstorm.cpp:159]: There's an enum for layers. It's in layer.hpp. You should use, use LAYER_LIGHTMAP.

File: src/video/color.cpp [hunk 2]
```diff
@@ -26,6 +26,7 @@ const Color Color::CYAN(0.0, 1.0, 1.0);
 const Color Color::MAGENTA(1.0, 0.0, 1.0);
 const Color Color::YELLOW(1.0, 1.0, 0.0);
 const Color Color::WHITE(1.0, 1.0, 1.0);
+const Color Color::LIGHTNING_HIT_COLOR(0.6f, 0.6f, 0.6f, 0.5f);
```
Comment (MatusGuy) [inline src/video/color.cpp:29]: why is this here??? This goes in thunderstorm.cpp

**Review by MatusGuy — CHANGES_REQUESTED** [review]

**Comment by tobbi** [comment]
> Personally I find the thunder too bright, making it uneasy on the eyes. Would there be an option to turn this off?

This was Rusty's idea. Battle with him.

**Comment by Rusty-Box** [comment]
It appears the white surface drawn when lightning strikes isn't even colored white. It looks rather gray. Even when the ambient light is set to r=1, g=1, b=1. Is its alpha value set to 1 when it flashes?

**Comment by Rusty-Box** [comment]
Ideally, if we get the flash to be white we may not even have to tamper with the ambient light at all during the flash as the blend mode "add" brightens it all up anyways (unless I am mistaking, though that's how it works with a pure white gradient set to "add" and a z-layer of 500).

**Comment by tobbi** [comment]
It starts at 1 and then quickly levels off to 0.

**Comment by Rusty-Box** [comment]
Huh, it doesn't look like it though at all. I did notice something else though too. You should remove the stuff related to `const Color Color::LIGHTNING_HIT_COLOR(0.6, 0.6, 0.6, 0.5);` because it actually makes a fully bright sector darker when lightning strikes

**Comment by Rusty-Box** [comment]
Not sure if that fixes all of it though since the white surfaces appears as not white even in a fully bright sector too.

**Comment by Rusty-Box** [comment]
This is how it is supposed to look like. To give you a visual reference!
_Note: I used a scripted Gradient object for this!_




https://github.com/SuperTux/supertux/assets/20615048/86e72405-97d5-4da0-95d9-184b5f44761a

**Review by MatusGuy — APPROVED** [review]

**Comment by tobbi** [comment]
https://github.com/SuperTux/supertux/assets/583169/ec82e7d8-983e-4ef2-ac3b-ce26f21b373f

**Comment by Rusty-Box** [comment]
I'm still convied that the blending mode didn't really apply. For comparison I set up a Gradient Object with the same values as the drawn lighning flash you added (and on the same z-layer -> 500). But as you can see it is not the same. I don't know why and rn my only assuption is that the code may not support blending modes for you methode of drawing it.

Gradient set to r:1, g:1, b:1, a:0.9 and blend mode "Additive":
screenshot000011

Lighning Flash in this PR, which is supposedly also "Additive":
screenshot000012

File: src/object/thunderstorm.cpp [hunk 3]
```diff
@@ -30,9 +32,8 @@
 namespace {
 
 const float LIGHTNING_DELAY = 2.0f;
-const float FLASH_DISPLAY_TIME = 0.1f;
+const float FLASH_DISPLAY_TIME = 2.0f;
```
Comment (MatusGuy) [inline src/object/thunderstorm.cpp:35]: two seconds? What did this mean again?

File: src/object/thunderstorm.hpp [hunk 4]
```diff
@@ -70,18 +70,22 @@ class Thunderstorm final : public GameObject,
...
-  
+
   std::string m_strike_script;
 
   Timer time_to_thunder; /**< counts down until next thunder */
   Timer time_to_lightning; /**< counts down until next lightning */
   Timer flash_display_timer; /**< counts down while flash is displayed */
 
   std::map<uint32_t, uint32_t> changing_tiles; /**< preserves the tiles which an electrocution should change */
+  Color flash_color;
```
Comment (MatusGuy) [inline src/object/thunderstorm.hpp:88]: add m_ prefix

**Review by MatusGuy — APPROVED** [review]
I think this code could be much better, but this works

**Comment by tobbi** [comment]
> I think this code could be much better, but this works

In what way could the code be better?

File: src/object/thunderstorm.cpp (hunk 3 above)
Comment (tobbi) [reply src/object/thunderstorm.cpp:35]: flash_display_timer; /**< counts down while flash is displayed */ So, yeah, that is the time that the flash is displayed.

**Comment by tobbi** [comment]
https://cdn.discordapp.com/attachments/790381979128168448/1213525190949077073/2024-03-02_17-35-20.mp4?ex=65f5ca67&is=65e35567&hm=a34f7ed66a5cd0bedd383441bfc2674334a9631fcbb9bcfa7d2d77d371618ea5&

**Review by Rusty-Box — APPROVED** [review]
There we go.

**Review by mrkubax10 — APPROVED** [review]

File: src/object/thunderstorm.cpp [hunk 5]
```diff
@@ -94,19 +96,35 @@ Thunderstorm::update(float )
...
 void
 Thunderstorm::draw(DrawingContext& context)
 {
   if (!flash_display_timer.started()) return;
 
-  float alpha = 0.33f;
   context.push_transform();
   context.set_translation(Vector(0, 0));
-  context.color().draw_filled_rect(context.get_rect(), Color(1, 1, 1, alpha), layer);
+  context.color().draw_gradient(m_flash_color, m_flash_color, 500, GradientDirection::HORIZONTAL, context.get_rect(), Blend::ADD);
```
Comment (Vankata453) [inline src/object/thunderstorm.cpp:126]: Before the gradient is drawn, the `DrawingContext` transform scale should be set to 1:

```
context.transform().scale = 1.f;
```

Doing so will prevent #2770.

File: src/object/thunderstorm.cpp [hunk 6]
```diff
@@ -94,19 +96,35 @@ Thunderstorm::update(float )
     lightning();
     time_to_thunder.start(interval);
   }
+
+  if(flash_display_timer.started()) {
```
Comment (Vankata453) [inline src/object/thunderstorm.cpp:100]: Per `CODINGSTYLE.md`:

```
if (foo)
{
  long();
  complex();
  series();
  of();
  calls();
}
```
(first brace on separate line)

File: src/video/canvas.cpp [hunk 7]
```diff
@@ -264,17 +264,19 @@ Canvas::draw_gradient(const Color& top, const Color& bottom, int layer,
 
 void
 Canvas::draw_filled_rect(const Rectf& rect, const Color& color,
-                         int layer)
+                         int layer, const Blend& blend)
```
Comment (Vankata453) [inline src/video/canvas.cpp:267]: This argument should be removed, since:

1. It's not passed, so it's unneeded.
2. Even if it was passed, it'd be better to use the other, more advanced `Canvas::draw_filled_rect()` overload for providing a blend.

**Review by Vankata453 — CHANGES_REQUESTED** [review]

File: src/object/thunderstorm.cpp [hunk 8]
```diff
@@ -18,6 +18,8 @@
 
 #include "audio/sound_manager.hpp"
 #include "editor/editor.hpp"
+#include "object/ambient_light.hpp"
```
Comment (Vankata453) [inline src/object/thunderstorm.cpp:21]: Is this include used?

Comment (tobbi) [reply src/object/thunderstorm.cpp:21]: Not anymore it ain't.

**Review by Vankata453 — APPROVED** [review]
Codewise looks good.
